File: apps/web/src/lib/api/seo-meta.ts
import type { SchemaType, SeoEntityType, SeoMetaRecord } from "@/lib/mock-data/seo";

export type ApiSeoMeta = {
  id: string;
  entity_type: string;
  title: string;
  url: string;
  meta_title: string;
  meta_description: string;
  og_image: string | null;
  canonical_url: string | null;
  indexable: boolean;
  schema_type: string;
  score: number;
  issues: string[];
};

export type ApiSeoMetaListResponse = {
  data: ApiSeoMeta[];
  meta: { count: number; avg_score: number };
};

export function apiSeoMetaToRecord(row: ApiSeoMeta): SeoMetaRecord {
  return {
    id: row.id,
    entityType: row.entity_type as SeoEntityType,
    title: row.title,
    url: row.url,
    metaTitle: row.meta_title,
    metaDescription: row.meta_description,
    ogImage: row.og_image ?? "",
    canonicalUrl: row.canonical_url ?? "",
    indexable: row.indexable,
    schemaType: row.schema_type as SchemaType,
    score: row.score,
    issues: row.issues ?? [],
  };
}

export function buildSeoMetaQuery(params?: { search?: string; entity_type?: string }): string {
  const q = new URLSearchParams();
  if (params?.search) q.set("search", params.search);
  if (params?.entity_type) q.set("entity_type", params.entity_type);
  const s = q.toString();
  return s ? `?${s}` : "";
}

File: apps/api/app/routers/seo_meta.py
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.deps.auth import require_write_access
from app.database import get_db
from app.models.seo_meta_record import SeoMetaRecord, load_json
from app.schemas.seo_meta import SeoMetaListMeta, SeoMetaListResponse, SeoMetaRead, SeoMetaResponse, SeoMetaUpdate

router = APIRouter(prefix="/meta", tags=["seo-meta"])


def _to_read(row: SeoMetaRecord) -> SeoMetaRead:
    return SeoMetaRead(
        id=row.id,
        company_id=row.company_id,
        entity_type=row.entity_type,
        title=row.title,
        url=row.url,
        meta_title=row.meta_title,
        meta_description=row.meta_description,
        og_image=row.og_image,
        canonical_url=row.canonical_url,
        indexable=row.indexable,
        schema_type=row.schema_type,
        score=row.score,
        issues=load_json(row.issues_json, []),
        created_at=row.created_at,
        updated_at=row.updated_at,
    )


@router.get("", response_model=SeoMetaListResponse)
def list_meta(
    db: Session = Depends(get_db),
    search: Optional[str] = Query(default=None),
    entity_type: Optional[str] = Query(default=None),
) -> SeoMetaListResponse:
    query = db.query(SeoMetaRecord).order_by(SeoMetaRecord.score.desc(), SeoMetaRecord.title)
    if entity_type:
        query = query.filter(SeoMetaRecord.entity_type == entity_type)
    if search:
        term = f"%{search.strip().lower()}%"
        query = query.filter(
            (SeoMetaRecord.title.ilike(term))
            | (SeoMetaRecord.url.ilike(term))
            | (SeoMetaRecord.meta_title.ilike(term)),
        )
    rows = query.all()
    data = [_to_read(r) for r in rows]
    avg = sum(r.score for r in rows) / len(rows) if rows else 0.0
    return SeoMetaListResponse(data=data, meta=SeoMetaListMeta(count=len(data), avg_score=avg))


@router.get("/{record_id}", response_model=SeoMetaResponse)
def get_meta(record_id: str, db: Session = Depends(get_db)) -> SeoMetaResponse:
    row = db.get(SeoMetaRecord, record_id)
    if not row:
        raise HTTPException(status_code=404, detail="Meta record not found")
    return SeoMetaResponse(data=_to_read(row))


@router.patch("/{record_id}", response_model=SeoMetaResponse)
def update_meta(
    record_id: str,
    payload: SeoMetaUpdate,
    db: Session = Depends(get_db),
    _: object = Depends(require_write_access),
) -> SeoMetaResponse:
    row = db.get(SeoMetaRecord, record_id)
    if not row:
        raise HTTPException(status_code=404, detail="Meta record not found")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(row, key, value)
    db.commit()
    db.refresh(row)
    return SeoMetaResponse(data=_to_read(row))
